src/server/service/diagnostics.rs

```diff
@@ -0,0 +1,682 @@
+use futures::Future;
+use grpcio::{RpcContext, UnarySink};
+use kvproto::diagnosticspb::{
+    Diagnostics, SearchLogRequest, SearchLogResponse, ServerInfoRequest, ServerInfoResponse,
+};
+
+use crate::server::Error;
+
+/// Service handles the RPC messages for the `Diagnostics` service.
+#[derive(Clone)]
+pub struct Service {
+    log_file: String,
+}
+
+impl Service {
+    pub fn new(log_file: String) -> Self {
+        Service { log_file }
+    }
+}
+
+impl Diagnostics for Service {
+    fn search_log(
+        &mut self,
+        ctx: RpcContext<'_>,
+        req: SearchLogRequest,
+        sink: UnarySink<SearchLogResponse>,
+    ) {
+        let future = log::search(&self.log_file, req)
+            .map_err(|err| Error::Other(box_err!("Search log error: {:?}", err)))
+            .and_then(|res| sink.success(res).map_err(Error::from))
+            .map_err(|e| debug!("Diagnostics rpc failed"; "err" => ?e));
+        ctx.spawn(future);
+    }
+
+    fn server_info(
+        &mut self,
+        _ctx: RpcContext<'_>,
+        _req: ServerInfoRequest,
+        _sink: UnarySink<ServerInfoResponse>,
+    ) {
+        unimplemented!()
+    }
+}
+
+mod log {
```

Comment:
Could you move it to a dedicated mod folder? Eg, `src/server/diagnostics/log.rs`.

Reply:
I think it's better to keep put the log module in this file (other services in the `service` module is a single file), but we can take this into account if someday the log sub module growth.